packages/legend-studio-preset-query-builder/src/models/protocols/pure/v1/V1_QueryBuilder_FunctionExpressionBuilder.ts

```diff
@@ -0,0 +1,427 @@
+          ),
+        );
+        variableExpression.genericType = precedingExpression.genericType;
+        processingContext.addInferredVariables(
+          variable.name,
+          variableExpression,
+        );
+      }
+    });
+  }
+  const processedParameters = [
+    precedingExpression,
+    parameters[1].accept_ValueSpecificationVisitor(
+      new V1_ValueSpecificationBuilder(
+        compileContext,
+        processingContext,
+        openVariables,
+      ),
+    ),
+  ];
+  const expression = buildBaseSimpleFunctionExpression(
+    processedParameters,
+    functionName,
+    compileContext,
+  );
+  return [expression, processedParameters];
+};
+
+export const V1_buildFilterFunctionExpression = (
+  functionName: string,
+  parameters: V1_ValueSpecification[],
+  openVariables: string[],
+  compileContext: V1_GraphBuilderContext,
+  processingContext: V1_ProcessingContext,
+): [SimpleFunctionExpression, ValueSpecification[]] => {
+  assertTrue(
+    parameters.length === 2,
+    `Can't build filter() expression: filter() expects 1 argument`,
+  );
+
+  const precedingExpression = parameters[0].accept_ValueSpecificationVisitor(
+    new V1_ValueSpecificationBuilder(
+      compileContext,
+      processingContext,
+      openVariables,
+    ),
+  );
+  if (precedingExpression.genericType) {
+    const lambda = parameters[1];
+    if (lambda instanceof V1_Lambda) {
+      lambda.parameters.forEach((variable): void => {
+        if (variable.name && !variable.class) {
+          const variableExpression = new VariableExpression(
+            variable.name,
+            precedingExpression.multiplicity,
+          );
+          variableExpression.genericType = precedingExpression.genericType;
+          processingContext.addInferredVariables(
+            variable.name,
+            variableExpression,
+          );
+        }
+      });
+    }
+  }
+  const processedParams = [
+    precedingExpression,
+    parameters[1].accept_ValueSpecificationVisitor(
+      new V1_ValueSpecificationBuilder(
+        compileContext,
+        processingContext,
+        openVariables,
+      ),
+    ),
+  ];
+  const expression = buildBaseSimpleFunctionExpression(
+    processedParams,
+    functionName,
+    compileContext,
+  );
+
+  // return type of filter() is the same as that of the function precedes it
+  expression.genericType = precedingExpression.genericType;
+  expression.multiplicity = precedingExpression.multiplicity;
+  return [expression, processedParams];
+};
+
+export const V1_buildProjectFunctionExpression = (
+  functionName: string,
+  parameters: V1_ValueSpecification[],
+  openVariables: string[],
+  compileContext: V1_GraphBuilderContext,
+  processingContext: V1_ProcessingContext,
+): [SimpleFunctionExpression, ValueSpecification[]] => {
+  assertTrue(
+    parameters.length === 2 || parameters.length === 3,
+    `Can't build project() expression: project() expects at 1-2 arguments`,
```

Comment:
2-3rguments?

Reply:
yes, there are 2 forms:

```
project([col(x|$x.name, 'name'])
project([x|$x.name], ['name'])
```

Reply:
yeah i meant your error message says 1-2 arguments, but should be 2-3 no?

Reply:
oh, because the first param is the preceding expression.